### DISCUSSION TITLE
Get DepotDownloader to use lancache

### DISCUSSION THREAD

**Discussion by nutterthanos** [SteamRE/DepotDownloader#248] [Q&A] [answered]
Hi i was wondering how you can get depotdownloader to use lancache?
if this already isn't implemented can this be added as a feature?
thanks

**Comment by nutterthanos** [comment]
ok so this get moved here and pretty much pushed aside from being answered?

**Reply by voided** [reply]
@nutterthanos Sorry for the lack of replies, we're a small team of volunteers. We generally move issues to discussions when they're not bug related.

However, I am interested in understanding your use case for lancache support if you're still looking for this functionality.

**Comment by GoTeamScotch** [comment]
DepotDownloader uses steam's servers to download game files, similar to the actual Steam client app. I would assume if your PC running DepotDownloader is setup to route through LanCache, then LanCache would serve back files same as usual. LanCache just detects if you're requesting files from a given domain (steam.com, etc) before it kicks in and serves files it has already downloaded. Does it seem to not work?

**Reply by crkinard** [reply]
Does not work. Watching DepotDownloader download Cyberpunk 2077. Resource Monitor is saying it's pulling from 151.205.4.161 (which obviously is not my local LanCache IP).

**Reply by nutterthanos** [reply]
It never worked for me hence why I asked this originally.

**Comment by LordRalex** [comment]
Steam doesn't just use the override of a DNS server to handle it. It does a lot more logic to decide if Lancaches are in use, and how to handle them. This is not logic that DepotDownloader does, and so it's only assuming it's the very large DNS overrides, which are not used.

https://github.com/lancachenet/monolithic/issues/85 has good information on this process.

In short:

> As of today's Steam Client update, there is now full support in Steam for a lancache. The way it works is by setting a DNS override for lancache.steamcontent.com pointing to the IP address of your cache server.
The Steam client will then use that cache server for all downloads. It will still make requests to the cache server using the hostname of the Valve content server to download from, but it will connect to your cache.
In addition, it will increase the maximum number of connections per IP from 4 to 32. This should solve the issue people have been seeing with slow initial downloads through lancache. It will also prevent the Steam Client from using Open Caching, which caused some downloads to not be cached correctly.

So, because DepotDownloader is not doing that same logic, it's not using the caches. These are the only DNS entries (the last one being the local CDNs to actually get said files)

media.steampowered.com
api.steampowered.com
cache6-dfw1.steamcontent.com

**Reply by voided** [reply]
@LordRalex could you provide a use-case for DD to utilize lancache?

DD is mostly designed to download data directly from content servers.

Are end users utilizing DD to download games in LAN environments?

**Reply by LordRalex** [reply]
I am a developer for a game server management panel. We use DepotDownloader to handle downloading games for each game server, as it's far easier to run and set up over trying to deal with Steamcmd and it's.... limitations.

Since each server should be independent as some might be on different versions or need different things, we don't use a single "location" to handle 1 game. If a game server is needed twice (let's use CSGO), then it needs to download it twice, which means going to Steam twice.

I would rather prefer administrators be able to use Lancache, which is the more accepted solution from Steam, to cache these downloads so that the panel is not being forced to go over the (slower) network each time.

This also makes my testing of software easier, as having to debug other issues with Steam stuff (like... Steam apparently marking everything as executable but DepotDownloader isn't... thanks Steam) much slower as to debug TF2, I'm being forced to download it each time. (This though is a limitation in my tester, but it works off a clean slate to do it's test)

In the older days, this was transparent enough to not need direct support, since it was done on the DNS level, and the cache domains were being handled there, so DepotDownloader (unintentionally) would use a lancache.

**Reply by yaakov-h** [reply]
Oh this is quite interesting.

With modern .NET now it should also be fairly simple to implement. I was looking into sending requests to a specific IP a while ago and it's fairly straightforward with [SocketsHttpHandler.ConnectCallback](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.socketshttphandler.connectcallback?view=net-8.0), no need to set custom hosts headers or deal with SNI.

How does Steam decide when to use a lancache server? Does it just do an A (or maybe AAAA?) lookup for `lancache.steamcontent.com`, and change its behaviour based on whether that returns NXDOMAIN or not? (And how does it handle multiple addresses in the response?)

**Reply by LordRalex** [reply]
I am not 100% sure how it fully decides it, so I'm taking what I've read from this issue where Lancache talks a bit about it: https://github.com/lancachenet/monolithic/issues/85

From what I've understood from the Lancache community (they know the best as they've talked with Steam)

1) Client does a resolve on lancache.steamcontent.com (I assume A record based on later rules). This already does exist, so it depends on how it resolves.
2) If the result is in the RFC1918 range, then a lancache is in use. Otherwise, it is not. `The requirements are that the address returned by the lancache.steamcontent.com lookup be on a private IP – so 10.0.0.0/8, 127.0.0.0/8, 172.16.0.0/12, and 192.168.0.0/16 for ipv4 or fc00::/7 and fe80::/10 for ipv6.` It doesn't state how it reacts to multiple IPs. I assume if 1 IP is in that range, it treats it as a lancache.
3) The Steam client will proxy all requests through that host using HTTP, with a “host” header indicating the upstream host it would have otherwise talked to. So, while requests are sent to lancache.steamcontent.com, they contain a HOST record that points to the upstream server (cache6-dfw1.steamcontent.com as an example).

I am not sure if they explicitly care about the HOST header being set correctly, but this is what they do currently. I assume it might be needed so that Lancache is able to connect to an upstream server to satisfy the request if it cannot. Since well, lancache.steamcontent.com I don't think it's actually designed to download such files. It would have to know that it actually needs to go to cache6-dfw1 to get the depot files if it can't satisfy it.

**Reply by tpill90** [reply]
@LordRalex Your understanding of how this all works is correct.  One thing I would like to point out is that the `HOST` header is 100% required in order for Lancache (Nginx) to properly download from the upstream Steam CDN, since `lancache.steamcontent.com` will point to a private IP.

@voided @yaakov-h I've already written this exact logic as part of [steam-lancache-prefill](https://github.com/tpill90/steam-lancache-prefill), which was designed specifically for Lancache.  Since this code is already written in C#, it may help speed things along.

There is a class specifically for detecting if a Lancache is somewhere on the network, see [LancacheIpResolver.cs](https://github.com/tpill90/lancache-prefill-common/blob/main/LancacheIpResolver.cs).  This could be a near drop in here, minus the `IAnsiConsole` part which may not be desired.  Example usage for resolving the IP of a Lancache instance would be : 
```csharp
// Ex, 192.168.1.99
var lancacheIpAddress = await LancacheIpResolver.ResolveLancacheIpAsync(AnsiConsole.Console, "lancache.steamcontent.com");
```

The remainder of the implementation would to just make a request to the resolved IP, and point to the real CDN using the `HOST` header.  I've done so [here](https://github.com/tpill90/steam-lancache-prefill/blob/master/SteamPrefill/Handlers/DownloadHandler.cs#L90), however I will say that it will likely be a bit more difficult to work this part into DD.  I was originally using SteamKit's chunk download method, but had to break it out into my own in order to be able to set this header.

Hope this can be of help to you all.

**Reply by Dimensional** [reply]
Looking through code a bit, I found that in CDNClientPool.cs, line 95 has an argument where the server type has to either be SteamCache or CDN. Perhaps this needs to be updated, since the SteamKit2 has this and Lancache listed in it's library?

**Reply by LordRalex** [reply]
One thing that I've noticed while also looking at it, is that HTTPS is also being used. This means that while the traffic can go through the lancache (even if you redirect all domains like before), because it's under HTTPS, it is still not able to cache the requests.

So, knowing that, the checks also would be "downgrading" the connection to the content server to HTTP if it's a cache server.

I don't even see a way to tell DepotDownloader to not use HTTPS, so there is no way to make it work as-is to use a cache server.

**Reply by nutterthanos** [reply]
currently there is not since there is no config file that you can mess with to get it to do that.
steamctl does but login system is old so i end up with http 429 or whatever on logins.
also steamctl is so broken it can't download any apps/games now
as i understand the list comes from https://api.steampowered.com/IContentServerDirectoryService/GetServersForSteamPipe/v1/?cell_id=
cell_id being After you log on, the CM sends you a "cell ID". Steam divides the world into regions, called "cells". [There's a list here.](https://github.com/SteamDatabase/SteamTracking/blob/master/ClientExtracted/steam/cached/CellMap.vdf) The official client saves this ID, and sends it back in subsequent login requests.

Edit: CellMap.vdf is gone
don't know how its determined anymore

**Reply by nutterthanos** [reply, marked as answer]
i got an working workaround
it involves using a proxy server like charles proxy and using like the rewrite function/map remote function to get it to run through the trigger domain lancache.steamcontent.com

This is temporary solution and not official but unofficially works

**Reply by Acrocanthosaur** [reply]
> i got an working workaround it involves using a proxy server like charles proxy and using like the rewrite function/map remote function to get it to run through the trigger domain lancache.steamcontent.com
> 
> This is temporary solution and not official but unofficially works

May i ask how u did it in details ?

**Reply by nutterthanos** [reply]
> May i ask how u did it in details ?

Do you already have Charles Proxy installed if not do you know where to find it and install it?

**Reply by Acrocanthosaur** [reply]
I do have it installed already

**Reply by nutterthanos** [reply]
> I do have it installed already

alright has it been configured already with ssl pinning?
if it has just go to Tools > Map Remote then click on add then add whatever is in these photos
![image](https://github.com/user-attachments/assets/7d30f02e-21a3-4a39-8577-aaaa6c863cdd)
![image](https://github.com/user-attachments/assets/92a1ca11-0bb4-4458-8cc1-3a7afc3b2567)
![image](https://github.com/user-attachments/assets/c7406443-eafa-47bf-b724-7df1e8911613)
![image](https://github.com/user-attachments/assets/d2ac78a3-bd60-4bb5-9dc0-9c0ae6f94d08)

**Comment by Kein** [comment]
Isnt it a .NET app? If you build non-single package you can use default config.json to specify dotnet proxy

**Comment by tpill90** [comment]
Lancache is now supported by DepotDownloader so this can be closed now

**Reply by nutterthanos** [reply]
technically since its an discussion it can't be closed unless its like limit to contributors or whatever it's called?

**Reply by tpill90** [reply]
They can be closed https://github.com/SteamRE/DepotDownloader/discussions?discussions_q=is%3Aclosed

**Comment by Kein** [comment]
A built-in proxy support would be nice too
